kornia/contrib/kmeans.py

```diff
@@ -0,0 +1,172 @@
+        self.cluster_centers = cluster_centers
+        self.tolerance = tolerance
+        self.max_iterations = max_iterations
+
+        self.final_cluster_assignments = None
+        self.final_cluster_centers = None
+
+        if seed is not None:
+            torch.manual_seed(seed)
+
+    def get_cluster_centers(self) -> Tensor:
+        KORNIA_CHECK(self.final_cluster_centers is not None, "Model has not been fit to a dataset")
+        return self.final_cluster_centers
+
+    def get_cluster_assignments(self) -> Tensor:
+        KORNIA_CHECK(self.final_cluster_assignments is not None, "Model has not been fit to a dataset")
+        return self.final_cluster_assignments
+
+    def _initialise_cluster_centers(self, X: Tensor, num_clusters: int) -> Tensor:
+        """Chooses num_cluster points from X as the initial cluster centers.
+
+        Args:
+            X: 2D input tensor to be clustered
+            num_clusters: number of desired cluster centers
+
+        Returns:
+            2D Tensor with num_cluster rows
+        """
+        num_samples: int = len(X)
+        perm = torch.randperm(num_samples)
+        idx = perm[:num_clusters]
+        initial_state = X[idx]
+        return initial_state
+
+    def _pairwise_euclidean_distance(self, data1: Tensor, data2: Tensor) -> Tensor:
+        """Computes pairwise squared distance between 2 sets of vectors.
+
+        Args:
+            data1: 2D tensor of shape N, D
+            data2: 2D tensor of shape C, D
+
+        Returns:
+            2D tensor of shape N, C
+        """
+        # N*1*D
+        A = data1[:, None, ...]
+        # 1*C*D
+        B = data2[None, ...]
+        distance = (A - B) ** 2.0
+        # return N*C matrix for pairwise distance
+        distance = distance.sum(dim=-1)
+        return distance
+
+    def fit(self, X: Tensor) -> None:
+        """Iterative KMeans clustering till a threshold for shift in cluster centers or a maximum no of iterations
+        have reached.
+
+        Args:
+            X: 2D input tensor to be clustered
+        """
+        # X should have only 2 dimensions
+        KORNIA_CHECK_SHAPE(X, ["N", "D"])
+
+        if self.cluster_centers is None:
+            self.cluster_centers = self._initialise_cluster_centers(X, self.num_clusters)
+        else:
+            # X and cluster_centers should have same number of columns
+            KORNIA_CHECK(
+                X.shape[1] == self.cluster_centers.shape[1],
+                f"Dimensions at position 1 of X and cluster_centers do not match. \
+                {X.shape[1]} != {self.cluster_centers.shape[1]}",
+            )
+
+        # X = X.to(self.device)
+        current_centers = self.cluster_centers
+
+        previous_centers: Tensor | None = None
+        iteration: int = 0
+
+        while True:
+            # find distance between X and current_centers
+            distance: Tensor = self._pairwise_euclidean_distance(X, current_centers)
```

Comment:
you can use existing `euclidean_distance`: https://github.com/kornia/kornia/blob/69a697020a5f63ba7d46369673709c4f23d72313/kornia/geometry/linalg.py#L241

Reply:
Hi line 254 does `KORNIA_CHECK(x.shape == y.shape)` which prevents me from using this function. As `len(X) != len(current_centers)`

Reply:
i think we can safely remove that check /cc @johnnv1